src/state.rs

```diff
@@ -106,6 +106,46 @@ impl State {
     pub(crate) fn get_context_u128(&self) -> u128 {
         self.current_frame.context_u128
     }
+
+    pub(crate) fn snapshot(&self) -> StateSnapshot {
+        assert!(self.heaps[self.current_frame.aux_heap].is_empty());
+        StateSnapshot {
+            registers: self.registers,
+            register_pointer_flags: self.register_pointer_flags,
+            flags: self.flags.clone(),
+            bootloader_frame: self.current_frame.snapshot(),
+            bootloader_heap_snapshot: self.heaps.snapshot(),
+            transaction_number: self.transaction_number,
+            context_u128: self.context_u128,
+        }
+    }
+
+    pub(crate) fn rollback(&mut self, snapshot: StateSnapshot) {
+        assert!(self.heaps[self.current_frame.aux_heap].is_empty());
+        let StateSnapshot {
+            registers,
+            register_pointer_flags,
+            flags,
+            bootloader_frame,
+            bootloader_heap_snapshot,
+            transaction_number,
+            context_u128,
+        } = snapshot;
+
+        for heap in self.current_frame.rollback(bootloader_frame) {
+            self.heaps.deallocate(heap);
+        }
```

Comment:
Just to understand better: Would encapsulating deallocation logic in the heap snapshot not work? E.g., snapshotting the number of heaps and deallocating heaps with larger IDs on the rollback. Could it be the case that the bootloader would keep heaps with smaller IDs alive, or that some of the heaps in the rolled back execution need to be retained anyway?

Reply:
All heaps created during the rolled back time are deallocated but most of them get deallocated earlier. Only the ones kept alive by the bootloader need to be cleaned up.

Heap numbering is not guaranteed to be sequential, though it currently is.